Introduce conversion functions for plugins

gassmoeller opened the pull request:
One of the features of ASPECT I understood the least in the beginning was the use of `dynamic_cast` to determine the actual type of an interface plugin pointer. This is a proposal to make this more explicit and easier to understand. There are two main advantages to the new functions:
- No `dynamic_cast` that confuses users, instead function names that describe what is done
- No use of null pointers anymore. Therefore, nobody can forget to check for a null pointer before using the object. Always return references.

If this is something that we would like to do, there are dozens of places over the whole codebase which can be simplified with the new functions.
Comments?

bangerth commented on include/aspect/postprocess/interface.h:
```diff
@@ -231,6 +232,28 @@ namespace aspect
+         */
+        template <typename PostprocessorType>
+        PostprocessorType &
+        get_postprocessor_of_type () const;
```
Would you want to deprecate `find_postprocess()` in that case?

Also, the docs still talk about pointers.

bangerth commented on include/aspect/postprocess/interface.h:
```diff
@@ -378,6 +401,46 @@ namespace aspect
+    {
+      AssertThrow(has_postprocessor_of_type<PostprocessorType> (),
+                  ExcMessage("You asked Postprocess:Manager::get_postprocessor_of_type() for a "
+                             "postprocessor type that could not be found in the current model. Activate this "
```
I think you can print the actual type of the requested postprocessor using `typeid(PostprocessorType).name()`. You'll have to demangle this, though. I think there is a function for this in `dealii::Utilities`.

bangerth commented on include/aspect/postprocess/interface.h:
```diff
@@ -378,6 +401,46 @@ namespace aspect
+           p = postprocessors.begin();
+           p != postprocessors.end(); ++p)
+        if (Utilities::object_is_of_type<PostprocessorType>(*(*p)))
+          postprocessor = p;
```
couldn't you just return right here? No need to keep the loop going.

bangerth commented on include/aspect/utilities.h:
```diff
@@ -1069,6 +1069,28 @@ namespace aspect
+    template <typename TestType, typename ObjectType>
+    inline
+    bool
+    object_is_of_type (const ObjectType &object)
```
I'm not fond of that name because the real object may actually be of type *derived* from `TestType`. But I also can't come up with a better name. Want to give it another try? Maybe `object_type_matches`?

gassmoeller commented:
I think I addressed your comments (I like the new name and the type information, and yes I intended to deprecate the `find_postprocessor` function), and added tests for the asserts. If you agree this is useful, should we merge this PR first, and I will create a follow-up that uses the new functions in all the other places?

bangerth commented on include/aspect/postprocess/interface.h:
```diff
@@ -226,10 +230,33 @@ namespace aspect
          * if one of them has the desired type specified by the template
          * argument. If so, return a pointer to it. If no postprocessor is
          * active that matches the given type, return a NULL pointer.
+         * @deprecated: Use has_matching_postprocessor() and
```
add an empty line here to make sure we don't actually wrap the whole paragraph. I don't think you need the colon.

bangerth commented on include/aspect/postprocess/interface.h:
```diff
@@ -226,10 +230,33 @@ namespace aspect
+        /**
+         * Go through the list of all postprocessors that have been selected
+         * in the input file (and are consequently currently active) and return
+         * if one of them has the desired type specified by the template
```
return *true*

bangerth commented on include/aspect/utilities.h:
```diff
@@ -1069,6 +1071,51 @@ namespace aspect
+                               + boost::core::demangle(typeid(PluginType).name())
+                               + "> into type <"
+                               + boost::core::demangle(typeid(TestType).name()) +
+                               "> in which it can not be converted."));
```
this reads awkwardly. say "...into type <...>. But this cast cannot be performed."

bangerth commented on include/aspect/utilities.h:
```diff
@@ -1069,6 +1071,51 @@ namespace aspect
+    get_plugin_as_type (PluginType &object)
+    {
+      if (!plugin_type_matches<TestType>(object))
+        AssertThrow(false,
```
Couldn't you move the condition into the assertion where you currently have `false`?

tjhei commented on include/aspect/postprocess/interface.h:
```diff
@@ -226,10 +230,34 @@ namespace aspect
+        /**
+         * Go through the list of all postprocessors that have been selected
+         * in the input file (and are consequently currently active) and see
+         * if one of them has the desired type specified by the template
```
" or can be casted to that type"? This might be important if we have additional base classes.

tjhei commented on include/aspect/utilities.h:
```diff
@@ -1069,6 +1071,50 @@ namespace aspect
+    bool
+    plugin_type_matches (const PluginType &object)
+    {
+      return (dynamic_cast<const TestType *> (&object) != 0);
```
NULL?

gassmoeller commented:
I addressed your comments, but have one more question. The `Utilities` namespace does not feel quite right for this type of functions, and it would require me to include utilities.h in a lot of new plugins. These functions provide functionality at the very core of the plugin architecture, so should they go into plugins.h? That header is included everywhere (every interface includes it). And should I put them into a new namespace `aspect::Plugins`? Because the only namespace in there is currently `aspect::internal::Plugins`, but these functions are not intended for internal use, they should be used as widely as possible.

gassmoeller commented:
I understood that I had the ok to merge with the modifications, so I am going ahead with that. Let me know if there is need for more discussions, since this is not yet used a lot we can still change things.
